instrumentation/lettuce/lettuce-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/lettuce/v4_0/LettuceAsyncClientTest.java

```diff
@@ -0,0 +1,568 @@
+                    span.hasName("GET")
+                        .hasKind(SpanKind.CLIENT)
+                        .hasParent(trace.getSpan(0))
+                        .hasAttributesSatisfyingExactly(
+                            equalTo(SemanticAttributes.DB_SYSTEM, "redis"),
+                            equalTo(SemanticAttributes.DB_OPERATION, "GET")),
+                span ->
+                    span.hasName("callback1")
+                        .hasKind(SpanKind.INTERNAL)
+                        .hasParent(trace.getSpan(0)),
+                span ->
+                    span.hasName("callback2")
+                        .hasKind(SpanKind.INTERNAL)
+                        .hasParent(trace.getSpan(0))));
+  }
+
+  @Test
+  void testCommandWithNoArgumentsUsingBiconsumer() {
+    AsyncConditions conditions = new AsyncConditions();
+    BiConsumer<String, Throwable> biConsumer =
+        (keyRetrieved, error) ->
+            testing.runWithSpan(
+                "callback", () -> conditions.evaluate(() -> assertThat(keyRetrieved).isNotNull()));
+
+    testing.runWithSpan(
+        "parent",
+        () -> {
+          RedisFuture<String> redisFuture = asyncCommands.randomkey();
+          redisFuture.whenCompleteAsync(biConsumer);
+        });
+
+    testing.waitAndAssertTraces(
+        trace ->
+            trace.hasSpansSatisfyingExactly(
+                span -> span.hasName("parent").hasKind(SpanKind.INTERNAL).hasNoParent(),
+                span ->
+                    span.hasName("RANDOMKEY")
+                        .hasKind(SpanKind.CLIENT)
+                        .hasParent(trace.getSpan(0))
+                        .hasAttributesSatisfyingExactly(
+                            equalTo(SemanticAttributes.DB_SYSTEM, "redis"),
+                            equalTo(SemanticAttributes.DB_OPERATION, "RANDOMKEY")),
+                span ->
+                    span.hasName("callback")
+                        .hasKind(SpanKind.INTERNAL)
+                        .hasParent(trace.getSpan(0))));
+  }
+
+  @Test
+  void testHashSetAndThenNestApplyToHashGetall() throws Throwable {
+    AsyncConditions conditions = new AsyncConditions();
+
+    RedisFuture<String> hmsetFuture = asyncCommands.hmset("TESTHM", testHashMap);
+    hmsetFuture.thenApplyAsync(
+        setResult -> {
+          // Wait for 'hmset' trace to get written
+          testing.waitForTraces(1);
+
+          conditions.evaluate(
+              () -> {
+                assertThat(setResult).isEqualTo("OK");
+              });
+          RedisFuture<Map<String, String>> hmGetAllFuture = asyncCommands.hgetall("TESTHM");
+          hmGetAllFuture.exceptionally(
+              error -> {
+                logger.error("unexpected: {}", error.toString());
+                return null;
+              });
+          hmGetAllFuture.thenAccept(
+              hmGetAllResult ->
+                  conditions.evaluate(() -> assertThat(testHashMap).isEqualTo(hmGetAllResult)));
```

Comment:
I think you could use a `CompletableFuture` for this entire block:

```suggestion
          if (!"OK".equals(setResult)) {
            future.completeExceptionally(new AssertionError("Wrong hmset result " + setResult));
            return null;
          }
          
          RedisFuture<Map<String, String>> hmGetAllFuture = asyncCommands.hgetall("TESTHM");
          hmGetAllFuture.whenComplete((result, exception) -> {
            if (exception != null) {
              future.completeExceptionally(exception);
            } else {
              future.complete(result);
            }
          });
```

And after that you can just `await()` on the `future` and assert on its content (or fail if there was an exception)